File: android/src/com/jamme/MobileKeycodes.java
package com.jamme;

import java.util.HashMap;
import java.util.Map;

import android.view.KeyEvent;

public final class MobileKeycodes {
	private MobileKeycodes() {
	}
	  
	private static final Map<Integer, Integer> mobileKeyMap = new HashMap<Integer, Integer>();
	  
	static {
		mapKey(KeyEvent.KEYCODE_VOLUME_UP, JAKeycodes.A_CAP_Q);
		mapKey(KeyEvent.KEYCODE_VOLUME_DOWN, JAKeycodes.A_CAP_E);
		mapKey(KeyEvent.KEYCODE_ESCAPE, JAKeycodes.A_ESCAPE);
		mapKey(KeyEvent.KEYCODE_BACK, JAKeycodes.A_ESCAPE);
		mapKey(KeyEvent.KEYCODE_MENU, JAKeycodes.A_CONSOLE);
		mapKey(KeyEvent.KEYCODE_TAB, JAKeycodes.A_TAB);
		mapKey(KeyEvent.KEYCODE_DPAD_CENTER, JAKeycodes.A_ENTER);
		mapKey(KeyEvent.KEYCODE_ENTER, JAKeycodes.A_ENTER);
		mapKey(KeyEvent.KEYCODE_SPACE, JAKeycodes.A_SPACE);
		mapKey(KeyEvent.KEYCODE_DEL, JAKeycodes.A_BACKSPACE);
		mapKey(KeyEvent.KEYCODE_DPAD_UP, JAKeycodes.A_CURSOR_UP);
		mapKey(KeyEvent.KEYCODE_DPAD_DOWN, JAKeycodes.A_CURSOR_DOWN);
		mapKey(KeyEvent.KEYCODE_DPAD_LEFT, JAKeycodes.A_CURSOR_LEFT);
		mapKey(KeyEvent.KEYCODE_DPAD_RIGHT, JAKeycodes.A_CURSOR_RIGHT);
		mapKey(KeyEvent.KEYCODE_ALT_LEFT, JAKeycodes.A_ALT);
		mapKey(KeyEvent.KEYCODE_ALT_RIGHT, JAKeycodes.A_ALT2);
		mapKey(KeyEvent.KEYCODE_CTRL_LEFT, JAKeycodes.A_CTRL);
		mapKey(KeyEvent.KEYCODE_CTRL_RIGHT, JAKeycodes.A_CTRL2);
		mapKey(KeyEvent.KEYCODE_SHIFT_LEFT, JAKeycodes.A_SHIFT);
		mapKey(KeyEvent.KEYCODE_SHIFT_RIGHT, JAKeycodes.A_SHIFT2);
		mapKey(KeyEvent.KEYCODE_F1, JAKeycodes.A_F1);
		mapKey(KeyEvent.KEYCODE_F2, JAKeycodes.A_F2);
		mapKey(KeyEvent.KEYCODE_F3, JAKeycodes.A_F3);
		mapKey(KeyEvent.KEYCODE_F4, JAKeycodes.A_F4);
		mapKey(KeyEvent.KEYCODE_F5, JAKeycodes.A_F5);
		mapKey(KeyEvent.KEYCODE_F6, JAKeycodes.A_F6);
		mapKey(KeyEvent.KEYCODE_F7, JAKeycodes.A_F7);
		mapKey(KeyEvent.KEYCODE_F8, JAKeycodes.A_F8);
		mapKey(KeyEvent.KEYCODE_F9, JAKeycodes.A_F9);
		mapKey(KeyEvent.KEYCODE_F10, JAKeycodes.A_F10);
		mapKey(KeyEvent.KEYCODE_F11, JAKeycodes.A_F11);
		mapKey(KeyEvent.KEYCODE_F12, JAKeycodes.A_F12);
		mapKey(KeyEvent.KEYCODE_FORWARD_DEL, JAKeycodes.A_BACKSPACE);
		mapKey(KeyEvent.KEYCODE_INSERT, JAKeycodes.A_INSERT);
		mapKey(KeyEvent.KEYCODE_PAGE_UP, JAKeycodes.A_PAGE_UP);
		mapKey(KeyEvent.KEYCODE_PAGE_DOWN, JAKeycodes.A_PAGE_DOWN);
		mapKey(KeyEvent.KEYCODE_MOVE_HOME, JAKeycodes.A_HOME);
		mapKey(KeyEvent.KEYCODE_MOVE_END, JAKeycodes.A_END);
		mapKey(KeyEvent.KEYCODE_BREAK, JAKeycodes.A_PAUSE);
	}
	  
	private static final void mapKey(int mobileCode, JAKeycodes jaCode) {
		mapKey(mobileCode, jaCode.ordinal());
	}
	  
	private static final void mapKey(int mobileCode, int jaCode) {
		mobileKeyMap.put(mobileCode, jaCode);
	}
	  
	public static int getMobileKey(int acode, int unicode) {
		final Integer result = mobileKeyMap.get(acode);
		return result != null ? result : (unicode < 128 ? Character.toLowerCase(unicode) : 0);
	}
}

File: android/Mono/MobileKeycodes.cs
using System.Collections.Generic;

using Android.Views;

namespace android {
	public static class MobileKeycodes {
		private static Dictionary<Keycode, JAKeycodes> MobileKeyMap = new Dictionary<Keycode, JAKeycodes>() {
			{Keycode.VolumeUp, JAKeycodes.A_CAP_Q},
			{Keycode.VolumeDown, JAKeycodes.A_CAP_E},
			{Keycode.Escape, JAKeycodes.A_ESCAPE},
			{Keycode.Back, JAKeycodes.A_ESCAPE},
			{Keycode.Menu, JAKeycodes.A_CONSOLE},
			{Keycode.Tab, JAKeycodes.A_TAB},
			{Keycode.DpadCenter, JAKeycodes.A_ENTER},
			{Keycode.Enter, JAKeycodes.A_ENTER},
			{Keycode.Space, JAKeycodes.A_SPACE},
			{Keycode.Del, JAKeycodes.A_BACKSPACE},
			{Keycode.DpadUp, JAKeycodes.A_CURSOR_UP},
			{Keycode.DpadDown, JAKeycodes.A_CURSOR_DOWN},
			{Keycode.DpadLeft, JAKeycodes.A_CURSOR_LEFT},
			{Keycode.DpadRight, JAKeycodes.A_CURSOR_RIGHT},
			{Keycode.AltLeft, JAKeycodes.A_ALT},
			{Keycode.AltRight, JAKeycodes.A_ALT2},
			{Keycode.CtrlLeft, JAKeycodes.A_CTRL},
			{Keycode.CtrlRight, JAKeycodes.A_CTRL2},
			{Keycode.ShiftLeft, JAKeycodes.A_SHIFT},
			{Keycode.ShiftRight, JAKeycodes.A_SHIFT2},
			{Keycode.F1, JAKeycodes.A_F1},
			{Keycode.F2, JAKeycodes.A_F2},
			{Keycode.F3, JAKeycodes.A_F3},
			{Keycode.F4, JAKeycodes.A_F4},
			{Keycode.F5, JAKeycodes.A_F5},
			{Keycode.F6, JAKeycodes.A_F6},
			{Keycode.F7, JAKeycodes.A_F7},
			{Keycode.F8, JAKeycodes.A_F8},
			{Keycode.F9, JAKeycodes.A_F9},
			{Keycode.F10, JAKeycodes.A_F10},
			{Keycode.F11, JAKeycodes.A_F11},
			{Keycode.F12, JAKeycodes.A_F12},
			{Keycode.ForwardDel, JAKeycodes.A_BACKSPACE},
			{Keycode.Insert, JAKeycodes.A_INSERT},
			{Keycode.PageUp, JAKeycodes.A_PAGE_UP},
			{Keycode.PageDown, JAKeycodes.A_PAGE_DOWN},
			{Keycode.MoveHome, JAKeycodes.A_HOME},
			{Keycode.MoveEnd, JAKeycodes.A_END},
			{Keycode.Break, JAKeycodes.A_PAUSE},
			{Keycode.A, JAKeycodes.A_LOW_A},
			{Keycode.B, JAKeycodes.A_LOW_B},
			{Keycode.C, JAKeycodes.A_LOW_C},
			{Keycode.D, JAKeycodes.A_LOW_D},
			{Keycode.E, JAKeycodes.A_LOW_E},
			{Keycode.F, JAKeycodes.A_LOW_F},
			{Keycode.G, JAKeycodes.A_LOW_G},
			{Keycode.H, JAKeycodes.A_LOW_H},
			{Keycode.I, JAKeycodes.A_LOW_I},
			{Keycode.J, JAKeycodes.A_LOW_J},
			{Keycode.K, JAKeycodes.A_LOW_K},
			{Keycode.L, JAKeycodes.A_LOW_L},
			{Keycode.M, JAKeycodes.A_LOW_M},
			{Keycode.N, JAKeycodes.A_LOW_N},
			{Keycode.O, JAKeycodes.A_LOW_O},
			{Keycode.P, JAKeycodes.A_LOW_P},
			{Keycode.Q, JAKeycodes.A_LOW_Q},
			{Keycode.R, JAKeycodes.A_LOW_R},
			{Keycode.S, JAKeycodes.A_LOW_S},
			{Keycode.T, JAKeycodes.A_LOW_T},
			{Keycode.U, JAKeycodes.A_LOW_U},
			{Keycode.V, JAKeycodes.A_LOW_V},
			{Keycode.W, JAKeycodes.A_LOW_W},
			{Keycode.X, JAKeycodes.A_LOW_X},
			{Keycode.Y, JAKeycodes.A_LOW_Y},
			{Keycode.Z, JAKeycodes.A_LOW_Z},
			{Keycode.CapsLock, JAKeycodes.A_CAPSLOCK},
			{Keycode.Num0, JAKeycodes.A_0},
			{Keycode.Num1, JAKeycodes.A_1},
			{Keycode.Num2, JAKeycodes.A_2},
			{Keycode.Num3, JAKeycodes.A_3},
			{Keycode.Num4, JAKeycodes.A_4},
			{Keycode.Num5, JAKeycodes.A_5},
			{Keycode.Num6, JAKeycodes.A_6},
			{Keycode.Num7, JAKeycodes.A_7},
			{Keycode.Num8, JAKeycodes.A_8},
			{Keycode.Num9, JAKeycodes.A_9},
			{Keycode.Minus, JAKeycodes.A_MINUS},
			{Keycode.Plus, JAKeycodes.A_PLUS},
			{Keycode.Period, JAKeycodes.A_PERIOD},
			{Keycode.Comma, JAKeycodes.A_COMMA},
			{Keycode.Semicolon, JAKeycodes.A_SEMICOLON},
			{Keycode.Apostrophe, JAKeycodes.A_SINGLE_QUOTE},
			{Keycode.Slash, JAKeycodes.A_FORWARD_SLASH},
			{Keycode.Backslash, JAKeycodes.A_BACKSLASH},
			{Keycode.LeftBracket, JAKeycodes.A_OPEN_SQUARE},
			{Keycode.RightBracket, JAKeycodes.A_CLOSE_SQUARE}
		};

		public static int GetMobileKey(Keycode acode, int unicode) {
			JAKeycodes result;
			try {
				result = MobileKeyMap[acode];
			} catch (KeyNotFoundException exception) {
				result = JAKeycodes.A_NULL;
			}
			return result != JAKeycodes.A_NULL ? (int)result : (unicode < 128 ? Java.Lang.Character.ToLowerCase(unicode) : 0);
		}
	}
}